### PR TITLE
feat(storage): Add support for multiple object storages

### PR DESCRIPTION
**Pull request by kmala**
Add support for multiple object storages

### CODE REVIEW

**Comment by kmala** [comment]
fixes #32

**Comment by smothiki** [comment]
Boto3 doesn't work for Ceph. Is it possible to move to boto 2.7

**Comment by arschles** [comment]
I've mentioned this before, but I think it got lost, so I'll repeat it here. In this repo, I see that you've reimplemented some of the same object storage code that is in `github.com/deis/distribution` (and used in https://github.com/deis/builder/pull/262) in python.

I get that this is a shell script component which can't directly access the various object storage systems, but IMO we shouldn't be re-implementing the object storage layer here when we already have logic for it in Go in that aforementioned repo.

My proposal is to build a simple command line tool (CLI) based on the github.com/deis/distribution codebase that can do the following:
- Read the appropriate environment variables for and object storage config file (based on the object storage system)
- Use the storage layer in https://github.com/deis/distribution to access the appropriate object storage system
- Do the operation requested (the operation would be specified via a command line argument)

Below is an example usage of the command I'm proposing:

``` console
objstorage download /tmp/slug.tgz --storage-type="$BUILDER_STORAGE"
```

In this example, the `objstorage` binary would know what files to read according to the given `--storage-type` flag, and as a result, what and how to connect to the given object storage system.

Thoughts?

**Comment by kmala** [comment]
If we see the code actually it is pretty similar to the thing what you proposed,just 5-10 line related to the download or upload and the remaining code is actually just getting the credentials from the environment.
I don't think there is any code  in here which is already present in the `github.com/deis/distribution` because here i just do a download or upload a file which is not present in the distribution repo where we have functions to read the contents or write the contents.

**Comment by smothiki** [comment]
I think the longterm solution for this would be to have a separate Cli repository like workflow-cli and then use the appropriate packages from that Cli repo rather than implementing code twice for same purpose.

**Comment by arschles** [comment]
The same logic to read the storage env var and the appropriate creds/config files is repeated in 3 repositories (`deis/dockerbuilder`, `deis/slugbuilder` and `deis/slugrunner`), and each repository pulls in the appropriate dependencies to access azure, s3 and gcs.

Currently we are reinventing the wheel in 3 times which I don't agree with. I've created https://github.com/deis/distribution/issues/8 to track this work.

**Comment by kmala** [comment]
If we want the CLI to be generic i think we shouldn't be reading the creds/config files in the CLI and rather pass it to the CLI as ENV variables or as a parameter or may be in another generic way.

**Comment by arschles** [comment]
@kmala please see https://github.com/deis/distribution/issues/8#issuecomment-202512697 for a specification of the CLI. I believe that this specification would make the first version of it fairly generic, with defaults that would make it relatively easy for us to begin using now.

We can then explore further changes to make it more generic in the future.

Please add your comments to that issue if you have them.

**Comment by smothiki** [comment]
Hope that Script exits with a non zero value when it fails . Code LGTM

**Comment by smothiki** [comment]
We could include a stat script here to check if upload is successful then we can avoid checking for object existence in builder. Script will do a non zero exit if stat of the object fails

**Comment by smothiki** [comment]
@kmala can you just modify the put object to include the above discussed functionality and remove stat object  from builder code

**Comment by smothiki** [comment]
Do we actually need the check to verify if an object exists I think we can completely avoid this check

**Comment by kmala** [comment]
@smothiki I don't see an issue with builder having the object stat code for now.Once the https://github.com/deis/distribution/issues/8 is done the CLI will take care of this scenario and we can remove the waitforobject code from everywhere

**Comment by smothiki** [comment]
Why dont we do that now for this PR

**Comment by smothiki** [comment]
Do we have to support a URL which doesn't require an auth

File: rootfs/bin/get_object [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  	# If you add port 80 to the end of the endpoint_url, boto3 freaks out.
+  	S3_URL = "http://"+mHost
+  else :
+  	S3_URL="http://"+mHost+":"+mPort
+
+  conn = boto3.resource('s3', aws_access_key_id=AWS_ACCESS_KEY_ID, aws_secret_access_key=AWS_SECRET_ACCESS_KEY, region_name=AWS_DEFAULT_REGION, endpoint_url=S3_URL, config=Config(signature_version='s3v4'))
+  # stop boto3 from automatically changing the endpoint
+  conn.meta.client.meta.events.unregister('before-sign.s3', fix_s3_host)
+
+  conn.Bucket(bucket_name).Object(tar_path).download_file('/tmp/slug.tgz')
```
Comment (arschles) [inline rootfs/bin/get_object:79]: Similar to what I commented at https://github.com/deis/slugrunner/pull/24#discussion_r57797483, can this get some unit tests?

Comment (arschles) [inline rootfs/bin/get_object:79]: Similar to https://github.com/deis/slugrunner/pull/24#discussion_r57812927, I am going to let this go and proceed with my LGTM.

**Pull request merged**